File: javascript/src/resources/volumes.ts
import type { HTTPClient } from "../internal/http.js";
import type { RequestOptions } from "../types/common.js";
import type {
  CreateVolumeParams,
  DownloadVolumeArchiveParams,
  GrepVolumeParams,
  ListVolumeEntriesParams,
  ListVolumeEntriesResponse,
  ListVolumesParams,
  ListVolumesResponse,
  PatchVolumeFileLinesParams,
  ReadVolumeFileLinesParams,
  ReadVolumeFileParams,
  ReadVolumeFileRawParams,
  SearchVolumeEntriesParams,
  SummarizeVolumeParams,
  UpdateVolumeParams,
  VolumeArchive,
  VolumeFileDeliver,
  VolumeFileLines,
  VolumeFileLinesPatch,
  VolumeFileRaw,
  VolumeFileWrite,
  VolumeGrepResponse,
  VolumeInfo,
  VolumePathDelete,
  VolumeSummary,
} from "../types/volumes.js";

export class VolumesResource {
  constructor(private readonly http: HTTPClient) {}

  list(params: ListVolumesParams = {}, options?: RequestOptions): Promise<ListVolumesResponse> {
    return this.http.request<ListVolumesResponse>(
      "GET",
      `/v1/volumes${buildSafetyQuery({ limit: params.limit, page_token: params.page_token, safety_identifier: params.safety_identifier })}`,
      undefined,
      options,
    );
  }

  create(params: CreateVolumeParams = {}, options?: RequestOptions): Promise<VolumeInfo> {
    return this.http.request<VolumeInfo>("POST", "/v1/volumes", params, options);
  }

  retrieve(volumeID: string, paramsOrOptions?: SafetyParams | RequestOptions, options?: RequestOptions): Promise<VolumeInfo> {
    const [params, requestOptions] = splitSafetyParams(paramsOrOptions, options);
    return this.http.request<VolumeInfo>(
      "GET",
      `/v1/volumes/${encodeURIComponent(volumeID)}${buildSafetyQuery({ safety_identifier: params.safety_identifier })}`,
      undefined,
      requestOptions,
    );
  }

  update(volumeID: string, params: UpdateVolumeParams, options?: RequestOptions): Promise<VolumeInfo> {
    const { safety_identifier, new_safety_identifier, ...body } = params;
    const requestBody: { name?: string; safety_identifier?: string } = body;
    if (new_safety_identifier !== undefined) {
      requestBody.safety_identifier = new_safety_identifier;
    }
    return this.http.request<VolumeInfo>(
      "PATCH",
      `/v1/volumes/${encodeURIComponent(volumeID)}${buildSafetyQuery({ safety_identifier }, new_safety_identifier !== undefined)}`,
      requestBody,
      options,
    );
  }

  delete(volumeID: string, paramsOrOptions?: SafetyParams | RequestOptions, options?: RequestOptions): Promise<void> {
    const [params, requestOptions] = splitSafetyParams(paramsOrOptions, options);
    return this.http.requestVoid(
      "DELETE",
      `/v1/volumes/${encodeURIComponent(volumeID)}${buildSafetyQuery({ safety_identifier: params.safety_identifier })}`,
      requestOptions,
    );
  }

  listEntries(
    volumeID: string,
    params: ListVolumeEntriesParams = {},
    options?: RequestOptions,
  ): Promise<ListVolumeEntriesResponse> {
    return this.http.request<ListVolumeEntriesResponse>(
      "GET",
      `/v1/volumes/${encodeURIComponent(volumeID)}/entries${buildSafetyQuery({
        path: params.path,
        limit: params.limit,
        page_token: params.page_token,
        safety_identifier: params.safety_identifier,
      })}`,
      undefined,
      options,
    );
  }

  searchEntries(
    volumeID: string,
    params: SearchVolumeEntriesParams,
    options?: RequestOptions,
  ): Promise<ListVolumeEntriesResponse> {
    return this.http.request<ListVolumeEntriesResponse>(
      "GET",
      `/v1/volumes/${encodeURIComponent(volumeID)}/search${buildSafetyQuery({
        query: params.query,
        path: params.path,
        limit: params.limit,
        page_token: params.page_token,
        safety_identifier: params.safety_identifier,
      })}`,
      undefined,
      options,
    );
  }

  readFile(
    volumeID: string,
    path: string,
    params: ReadVolumeFileRawParams,
    options?: RequestOptions,
  ): Promise<VolumeFileRaw>;
  readFile(
    volumeID: string,
    path: string,
    params?: ReadVolumeFileParams,
    options?: RequestOptions,
  ): Promise<VolumeFileDeliver>;
  readFile(
    volumeID: string,
    path: string,
    params: ReadVolumeFileParams | ReadVolumeFileRawParams = {},
    options?: RequestOptions,
  ): Promise<VolumeFileDeliver | VolumeFileRaw> {
    const format = "format" in params ? params.format : undefined;
    const url = `/v1/volumes/${encodeURIComponent(volumeID)}/files/${volumePath(path)}${buildSafetyQuery({
      max_bytes: params.max_bytes,
      format,
      safety_identifier: params.safety_identifier,
    })}`;
    if (format === "raw") {
      return this.http.requestBinary("GET", url, options).then(({ body, headers }) => ({
        path,
        size: Number(headers.get("X-Volume-Size") ?? body.byteLength),
        truncated: headers.get("X-Volume-Truncated") === "true",
        content: body,
        content_type: headers.get("Content-Type") ?? undefined,
      }));
    }
    return this.http.request<VolumeFileDeliver>("GET", url, undefined, options);
  }

  writeFile(
    volumeID: string,
    path: string,
    content: string | ArrayBuffer | Blob,
    paramsOrOptions?: SafetyParams | RequestOptions,
    options?: RequestOptions,
  ): Promise<VolumeFileWrite> {
    const [params, requestOptions] = splitSafetyParams(paramsOrOptions, options);
    return this.http.requestRaw<VolumeFileWrite>(
      "PUT",
      `/v1/volumes/${encodeURIComponent(volumeID)}/files/${volumePath(path)}${buildSafetyQuery({ safety_identifier: params.safety_identifier })}`,
      content,
      requestOptions,
    );
  }

  deletePath(volumeID: string, path: string, paramsOrOptions?: SafetyParams | RequestOptions, options?: RequestOptions): Promise<VolumePathDelete> {
    const [params, requestOptions] = splitSafetyParams(paramsOrOptions, options);
    return this.http.request<VolumePathDelete>(
      "DELETE",
      `/v1/volumes/${encodeURIComponent(volumeID)}/paths/${volumePath(path)}${buildSafetyQuery({ safety_identifier: params.safety_identifier })}`,
      undefined,
      requestOptions,
    );
  }

  reconcileUsage(volumeID: string, paramsOrOptions?: SafetyParams | RequestOptions, options?: RequestOptions): Promise<VolumeInfo> {
    const [params, requestOptions] = splitSafetyParams(paramsOrOptions, options);
    return this.http.request<VolumeInfo>(
      "POST",
      `/v1/volumes/${encodeURIComponent(volumeID)}/usage/reconcile${buildSafetyQuery({ safety_identifier: params.safety_identifier })}`,
      undefined,
      requestOptions,
    );
  }

  createDirectory(volumeID: string, path: string, paramsOrOptions?: SafetyParams | RequestOptions, options?: RequestOptions): Promise<{ path: string }> {
    const [params, requestOptions] = splitSafetyParams(paramsOrOptions, options);
    return this.http.request<{ path: string }>(
      "POST",
      `/v1/volumes/${encodeURIComponent(volumeID)}/directories${buildSafetyQuery({ safety_identifier: params.safety_identifier })}`,
      { path },
      requestOptions,
    );
  }

  downloadArchive(
    volumeID: string,
    params: DownloadVolumeArchiveParams = {},
    options?: RequestOptions,
  ): Promise<VolumeArchive> {
    const archivePath = normalizeArchivePath(params.path);
    return this.http
      .requestBinary(
        "GET",
        `/v1/volumes/${encodeURIComponent(volumeID)}/archive${buildSafetyQuery({ path: archivePath || undefined, safety_identifier: params.safety_identifier })}`,
        options,
      )
      .then(({ body, headers }) => ({
        path: archivePath,
        content: body,
        content_type: headers.get("Content-Type") ?? undefined,
      }));
  }

  summarize(volumeID: string, params: SummarizeVolumeParams = {}, options?: RequestOptions): Promise<VolumeSummary> {
    return this.http.request<VolumeSummary>(
      "POST",
      `/v1/volumes/${encodeURIComponent(volumeID)}/summarize${buildSafetyQuery({ safety_identifier: params.safety_identifier })}`,
      { path: params.path },
      options,
    );
  }

  readLines(
    volumeID: string,
    path: string,
    params: ReadVolumeFileLinesParams,
    options?: RequestOptions,
  ): Promise<VolumeFileLines> {
    return this.http.request<VolumeFileLines>(
      "GET",
      `/v1/volumes/${encodeURIComponent(volumeID)}/file_lines/${volumePath(path)}${buildSafetyQuery({
        start_line: params.start_line,
        end_line: params.end_line,
        max_bytes: params.max_bytes,
        safety_identifier: params.safety_identifier,
      })}`,
      undefined,
      options,
    );
  }

  patchLines(
    volumeID: string,
    path: string,
    params: PatchVolumeFileLinesParams,
    options?: RequestOptions,
  ): Promise<VolumeFileLinesPatch> {
    return this.http.request<VolumeFileLinesPatch>(
      "PATCH",
      `/v1/volumes/${encodeURIComponent(volumeID)}/file_lines/${volumePath(path)}${buildSafetyQuery({ safety_identifier: params.safety_identifier })}`,
      { start_line: params.start_line, end_line: params.end_line, replacement: params.replacement },
      options,
    );
  }

  grep(volumeID: string, params: GrepVolumeParams, options?: RequestOptions): Promise<VolumeGrepResponse> {
    return this.http.request<VolumeGrepResponse>(
      "GET",
      `/v1/volumes/${encodeURIComponent(volumeID)}/grep${buildSafetyQuery({
        pattern: params.pattern,
        path: params.path,
        limit: params.limit,
        page_token: params.page_token,
        safety_identifier: params.safety_identifier,
      })}`,
      undefined,
      options,
    );
  }
}

type SafetyParams = { safety_identifier?: string };

function splitSafetyParams(paramsOrOptions?: SafetyParams | RequestOptions, options?: RequestOptions): [SafetyParams, RequestOptions | undefined] {
  if (!paramsOrOptions) {
    return [{}, options];
  }
  if (
    "headers" in paramsOrOptions ||
    "timeout" in paramsOrOptions ||
    "signal" in paramsOrOptions ||
    "maxRetries" in paramsOrOptions
  ) {
    return [{}, paramsOrOptions];
  }
  return [paramsOrOptions as SafetyParams, options];
}

function buildSafetyQuery(values: Record<string, string | number | undefined>, forceSafety = false): string {
  const search = new URLSearchParams();
  for (const [key, value] of Object.entries(values)) {
    if (value !== undefined && (value !== "" || (forceSafety && key === "safety_identifier"))) {
      search.set(key, String(value));
    }
  }
  const qs = search.toString();
  return qs ? `?${qs}` : "";
}

function volumePath(path: string): string {
  return path
    .split("/")
    .filter(Boolean)
    .map((part) => encodeURIComponent(part))
    .join("/");
}

function normalizeArchivePath(path?: string): string {
  if (!path) {
    return "";
  }
  return path.trim().replace(/^\/+/, "").replace(/\/+/g, "/").replace(/\/$/, "");
}

File: python/src/agent_api/resources/volumes.py
from __future__ import annotations

from typing import Any, Literal, overload
from urllib.parse import quote, urlencode

from agent_api._utils import build_query, drop_none
from agent_api.types.volumes import (
    ListVolumeEntriesResponse,
    ListVolumesResponse,
    Volume,
    VolumeArchive,
    VolumeFileDeliver,
    VolumeFileLines,
    VolumeFileLinesPatch,
    VolumeFileRaw,
    VolumeFileWrite,
    VolumeGrepResponse,
    VolumePathDelete,
    VolumeSummary,
)


class VolumesAPI:
    def __init__(self, http: Any) -> None:
        self._http = http

    def list(
        self,
        *,
        limit: int | None = None,
        page_token: str | None = None,
        safety_identifier: str | None = None,
        user_id: str | None = None,
    ) -> ListVolumesResponse:
        return self._http.request(
            "GET",
            "/v1/volumes"
            + _safety_query(
                {"limit": limit, "page_token": page_token, "safety_identifier": safety_identifier, "user_id": user_id}
            ),
            None,
        )

    def create(self, *, name: str | None = None, safety_identifier: str | None = None) -> Volume:
        return self._http.request("POST", "/v1/volumes", drop_none({"name": name, "safety_identifier": safety_identifier}))

    def retrieve(self, volume_id: str, *, safety_identifier: str | None = None) -> Volume:
        return self._http.request("GET", f"/v1/volumes/{quote(volume_id, safe='')}" + _safety_query({"safety_identifier": safety_identifier}), None)

    def update(
        self,
        volume_id: str,
        *,
        name: str | None = None,
        safety_identifier: str | None = None,
        new_safety_identifier: str | None = None,
    ) -> Volume:
        body = drop_none({"name": name})
        if new_safety_identifier is not None:
            body["safety_identifier"] = new_safety_identifier
        return self._http.request(
            "PATCH",
            f"/v1/volumes/{quote(volume_id, safe='')}" + _safety_query({"safety_identifier": safety_identifier}, force_safety=new_safety_identifier is not None),
            body,
        )

    def delete(self, volume_id: str, *, safety_identifier: str | None = None) -> None:
        self._http.request_void("DELETE", f"/v1/volumes/{quote(volume_id, safe='')}" + _safety_query({"safety_identifier": safety_identifier}))

    def list_entries(
        self,
        volume_id: str,
        *,
        path: str | None = None,
        limit: int | None = None,
        page_token: str | None = None,
    ) -> ListVolumeEntriesResponse:
        return self._http.request(
            "GET",
            f"/v1/volumes/{quote(volume_id, safe='')}/entries"
            + build_query({"path": path, "limit": limit, "page_token": page_token}),
            None,
        )

    def search_entries(
        self,
        volume_id: str,
        *,
        query: str,
        path: str | None = None,
        limit: int | None = None,
        page_token: str | None = None,
    ) -> ListVolumeEntriesResponse:
        return self._http.request(
            "GET",
            f"/v1/volumes/{quote(volume_id, safe='')}/search"
            + build_query({"query": query, "path": path, "limit": limit, "page_token": page_token}),
            None,
        )

    @overload
    def read_file(
        self,
        volume_id: str,
        path: str,
        *,
        format: Literal["raw"],
        max_bytes: int | None = None,
    ) -> VolumeFileRaw: ...

    @overload
    def read_file(
        self,
        volume_id: str,
        path: str,
        *,
        max_bytes: int | None = None,
        format: None = None,
    ) -> VolumeFileDeliver: ...

    def read_file(
        self,
        volume_id: str,
        path: str,
        *,
        max_bytes: int | None = None,
        format: Literal["raw"] | None = None,
    ) -> VolumeFileDeliver | VolumeFileRaw:
        url = (
            f"/v1/volumes/{quote(volume_id, safe='')}/files/{_volume_path(path)}"
            + build_query({"max_bytes": max_bytes, "format": format})
        )
        if format == "raw":
            content, headers = self._http.request_binary("GET", url)
            size_header = headers.get("X-Volume-Size")
            return {
                "path": path,
                "size": int(size_header) if size_header else len(content),
                "truncated": headers.get("X-Volume-Truncated") == "true",
                "content": content,
                "content_type": headers.get("content-type"),
            }
        return self._http.request("GET", url, None)

    def write_file(self, volume_id: str, path: str, content: bytes | str) -> VolumeFileWrite:
        return self._http.request_raw("PUT", f"/v1/volumes/{quote(volume_id, safe='')}/files/{_volume_path(path)}", content)

    def delete_path(self, volume_id: str, path: str) -> VolumePathDelete:
        return self._http.request("DELETE", f"/v1/volumes/{quote(volume_id, safe='')}/paths/{_volume_path(path)}", None)

    def reconcile_usage(self, volume_id: str) -> Volume:
        return self._http.request(
            "POST",
            f"/v1/volumes/{quote(volume_id, safe='')}/usage/reconcile",
            None,
        )

    def create_directory(self, volume_id: str, path: str) -> dict[str, str]:
        return self._http.request(
            "POST",
            f"/v1/volumes/{quote(volume_id, safe='')}/directories",
            {"path": path},
        )

    def download_archive(self, volume_id: str, *, path: str | None = None) -> VolumeArchive:
        normalized = _normalize_archive_path(path)
        url = (
            f"/v1/volumes/{quote(volume_id, safe='')}/archive"
            + build_query({"path": normalized or None})
        )
        content, headers = self._http.request_binary("GET", url)
        return {
            "path": normalized,
            "content": content,
            "content_type": headers.get("content-type"),
        }

    def summarize(self, volume_id: str, *, path: str | None = None) -> VolumeSummary:
        return self._http.request(
            "POST",
            f"/v1/volumes/{quote(volume_id, safe='')}/summarize",
            drop_none({"path": path}),
        )

    def read_lines(
        self,
        volume_id: str,
        path: str,
        *,
        start_line: int,
        end_line: int | None = None,
        max_bytes: int | None = None,
    ) -> VolumeFileLines:
        return self._http.request(
            "GET",
            f"/v1/volumes/{quote(volume_id, safe='')}/file_lines/{_volume_path(path)}"
            + build_query({"start_line": start_line, "end_line": end_line, "max_bytes": max_bytes}),
            None,
        )

    def patch_lines(
        self,
        volume_id: str,
        path: str,
        *,
        start_line: int,
        end_line: int | None = None,
        replacement: str = "",
    ) -> VolumeFileLinesPatch:
        return self._http.request(
            "PATCH",
            f"/v1/volumes/{quote(volume_id, safe='')}/file_lines/{_volume_path(path)}",
            drop_none({"start_line": start_line, "end_line": end_line, "replacement": replacement}),
        )

    def grep(
        self,
        volume_id: str,
        *,
        pattern: str,
        path: str | None = None,
        limit: int | None = None,
        page_token: str | None = None,
    ) -> VolumeGrepResponse:
        return self._http.request(
            "GET",
            f"/v1/volumes/{quote(volume_id, safe='')}/grep"
            + build_query({"pattern": pattern, "path": path, "limit": limit, "page_token": page_token}),
            None,
        )


class AsyncVolumesAPI:
    def __init__(self, http: Any) -> None:
        self._http = http

    async def list(
        self,
        *,
        limit: int | None = None,
        page_token: str | None = None,
        safety_identifier: str | None = None,
        user_id: str | None = None,
    ) -> ListVolumesResponse:
        return await self._http.request(
            "GET",
            "/v1/volumes"
            + _safety_query(
                {"limit": limit, "page_token": page_token, "safety_identifier": safety_identifier, "user_id": user_id}
            ),
            None,
        )

    async def create(self, *, name: str | None = None, safety_identifier: str | None = None) -> Volume:
        return await self._http.request("POST", "/v1/volumes", drop_none({"name": name, "safety_identifier": safety_identifier}))

    async def retrieve(self, volume_id: str, *, safety_identifier: str | None = None) -> Volume:
        return await self._http.request("GET", f"/v1/volumes/{quote(volume_id, safe='')}" + _safety_query({"safety_identifier": safety_identifier}), None)

    async def update(
        self,
        volume_id: str,
        *,
        name: str | None = None,
        safety_identifier: str | None = None,
        new_safety_identifier: str | None = None,
    ) -> Volume:
        body = drop_none({"name": name})
        if new_safety_identifier is not None:
            body["safety_identifier"] = new_safety_identifier
        return await self._http.request(
            "PATCH",
            f"/v1/volumes/{quote(volume_id, safe='')}" + _safety_query({"safety_identifier": safety_identifier}, force_safety=new_safety_identifier is not None),
            body,
        )

    async def delete(self, volume_id: str, *, safety_identifier: str | None = None) -> None:
        await self._http.request_void("DELETE", f"/v1/volumes/{quote(volume_id, safe='')}" + _safety_query({"safety_identifier": safety_identifier}))

    async def list_entries(
        self,
        volume_id: str,
        *,
        path: str | None = None,
        limit: int | None = None,
        page_token: str | None = None,
    ) -> ListVolumeEntriesResponse:
        return await self._http.request(
            "GET",
            f"/v1/volumes/{quote(volume_id, safe='')}/entries"
            + build_query({"path": path, "limit": limit, "page_token": page_token}),
            None,
        )

    async def search_entries(
        self,
        volume_id: str,
        *,
        query: str,
        path: str | None = None,
        limit: int | None = None,
        page_token: str | None = None,
    ) -> ListVolumeEntriesResponse:
        return await self._http.request(
            "GET",
            f"/v1/volumes/{quote(volume_id, safe='')}/search"
            + build_query({"query": query, "path": path, "limit": limit, "page_token": page_token}),
            None,
        )

    @overload
    async def read_file(
        self,
        volume_id: str,
        path: str,
        *,
        format: Literal["raw"],
        max_bytes: int | None = None,
    ) -> VolumeFileRaw: ...

    @overload
    async def read_file(
        self,
        volume_id: str,
        path: str,
        *,
        max_bytes: int | None = None,
        format: None = None,
    ) -> VolumeFileDeliver: ...

    async def read_file(
        self,
        volume_id: str,
        path: str,
        *,
        max_bytes: int | None = None,
        format: Literal["raw"] | None = None,
    ) -> VolumeFileDeliver | VolumeFileRaw:
        url = (
            f"/v1/volumes/{quote(volume_id, safe='')}/files/{_volume_path(path)}"
            + build_query({"max_bytes": max_bytes, "format": format})
        )
        if format == "raw":
            content, headers = await self._http.request_binary("GET", url)
            size_header = headers.get("X-Volume-Size")
            return {
                "path": path,
                "size": int(size_header) if size_header else len(content),
                "truncated": headers.get("X-Volume-Truncated") == "true",
                "content": content,
                "content_type": headers.get("content-type"),
            }
        return await self._http.request("GET", url, None)

    async def write_file(self, volume_id: str, path: str, content: bytes | str) -> VolumeFileWrite:
        return await self._http.request_raw(
            "PUT",
            f"/v1/volumes/{quote(volume_id, safe='')}/files/{_volume_path(path)}",
            content,
        )

    async def delete_path(self, volume_id: str, path: str) -> VolumePathDelete:
        return await self._http.request(
            "DELETE",
            f"/v1/volumes/{quote(volume_id, safe='')}/paths/{_volume_path(path)}",
            None,
        )

    async def reconcile_usage(self, volume_id: str) -> Volume:
        return await self._http.request(
            "POST",
            f"/v1/volumes/{quote(volume_id, safe='')}/usage/reconcile",
            None,
        )

    async def create_directory(self, volume_id: str, path: str) -> dict[str, str]:
        return await self._http.request(
            "POST",
            f"/v1/volumes/{quote(volume_id, safe='')}/directories",
            {"path": path},
        )

    async def download_archive(self, volume_id: str, *, path: str | None = None) -> VolumeArchive:
        normalized = _normalize_archive_path(path)
        url = (
            f"/v1/volumes/{quote(volume_id, safe='')}/archive"
            + build_query({"path": normalized or None})
        )
        content, headers = await self._http.request_binary("GET", url)
        return {
            "path": normalized,
            "content": content,
            "content_type": headers.get("content-type"),
        }

    async def summarize(self, volume_id: str, *, path: str | None = None) -> VolumeSummary:
        return await self._http.request(
            "POST",
            f"/v1/volumes/{quote(volume_id, safe='')}/summarize",
            drop_none({"path": path}),
        )

    async def read_lines(
        self,
        volume_id: str,
        path: str,
        *,
        start_line: int,
        end_line: int | None = None,
        max_bytes: int | None = None,
    ) -> VolumeFileLines:
        return await self._http.request(
            "GET",
            f"/v1/volumes/{quote(volume_id, safe='')}/file_lines/{_volume_path(path)}"
            + build_query({"start_line": start_line, "end_line": end_line, "max_bytes": max_bytes}),
            None,
        )

    async def patch_lines(
        self,
        volume_id: str,
        path: str,
        *,
        start_line: int,
        end_line: int | None = None,
        replacement: str = "",
    ) -> VolumeFileLinesPatch:
        return await self._http.request(
            "PATCH",
            f"/v1/volumes/{quote(volume_id, safe='')}/file_lines/{_volume_path(path)}",
            drop_none({"start_line": start_line, "end_line": end_line, "replacement": replacement}),
        )

    async def grep(
        self,
        volume_id: str,
        *,
        pattern: str,
        path: str | None = None,
        limit: int | None = None,
        page_token: str | None = None,
    ) -> VolumeGrepResponse:
        return await self._http.request(
            "GET",
            f"/v1/volumes/{quote(volume_id, safe='')}/grep"
            + build_query({"pattern": pattern, "path": path, "limit": limit, "page_token": page_token}),
            None,
        )


def _normalize_archive_path(path: str | None) -> str:
    if not path:
        return ""
    return path.strip().strip("/").replace("//", "/")


def _volume_path(path: str) -> str:
    return "/".join(quote(part, safe="") for part in path.split("/") if part)


def _safety_query(params: dict[str, Any], *, force_safety: bool = False) -> str:
    filtered = {
        key: value
        for key, value in params.items()
        if value is not None and (value != "" or (force_safety and key == "safety_identifier"))
    }
    return "?" + urlencode(filtered) if filtered else ""
